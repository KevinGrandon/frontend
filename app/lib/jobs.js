// @flow

import JobStates from '../constants/JobStates';

export function jobTime(job: Object) {
  const { state, startedAt, finishedAt, canceledAt, timedOutAt } = job;
  const time = {};

  // job never stared, so no duration to show
  if (!startedAt) {
    return time;
  }

  time.from = startedAt;

  switch (state) {
    case JobStates.FINISHED:
      time.to = finishedAt;
      break;

    case JobStates.CANCELED:
    case JobStates.CANCELING:
      time.to = canceledAt;
      break;

    case JobStates.TIMED_OUT:
    case JobStates.TIMING_OUT:
      time.to = timedOutAt;
      break;
  }

  return time;
}

export function jobWaitTime(job: Object) {
  const { state, scheduledAt, startedAt } = job;
  const time = {};
  
  // job is blocked, don't bother to calculate
  if (state === JobStates.BLOCKED) {
    return time;
  }

  time.from = scheduledAt;

  if (startedAt) {
    time.to = startedAt;
  }
  
  return time;
}
