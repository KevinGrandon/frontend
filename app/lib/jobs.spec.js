/* global describe, it, expect */
import { jobTime, jobWaitTime } from './jobs';

const JOB_STATES = [
  'PENDING',
  'WAITING',
  'WAITING_FAILED',
  'BLOCKED',
  'BLOCKED_FAILED',
  'UNBLOCKED',
  'UNBLOCKED_FAILED',
  'SCHEDULED',
  'ASSIGNED',
  'ACCEPTED',
  'RUNNING',
  'FINISHED',
  'CANCELING',
  'CANCELED',
  'TIMING_OUT',
  'TIMED_OUT',
  'SKIPPED',
  'BROKEN'
];

describe('jobTime', () => {
  JOB_STATES.forEach((state) => {
    it(`returns correct time properties for the \`${state}\` state`, () => {
      expect(jobTime({
        state,
        canceledAt: new Date(1475643467000),
        startedAt: new Date(1475643466920),
        finishedAt: new Date(1475643467000),
        timedOutAt: new Date(1475644466920)
      })).toMatchSnapshot();

      expect(jobTime({
        state,
        canceledAt: new Date(1475643467000),
        finishedAt: new Date(1475643467000)
      })).toMatchSnapshot();

      expect(jobTime({
        state,
        timedOutAt: new Date(1475644466920),
        canceledAt: new Date(1475643467000),
        finishedAt: new Date(1475643467000)
      })).toMatchSnapshot();
    });
  });
});

describe('jobWaitTime', () => {
  JOB_STATES.forEach((state) => {
    it(`returns correct time properties for the \`${state}\` state`, () => {
      expect(jobWaitTime({
        state,
        scheduledAt: "2017-10-29T00:32:23.479Z",
        startedAt: "2017-10-29T00:34:37.000Z"
      })).toMatchSnapshot();

      expect(jobWaitTime({
        state,
        scheduledAt: "2017-10-29T00:32:23.479Z"
      })).toMatchSnapshot();
    });
  });
});
