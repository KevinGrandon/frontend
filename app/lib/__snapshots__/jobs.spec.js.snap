// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`jobTime returns correct time properties for the \`ACCEPTED\` state 1`] = `
Object {
  "from": 2016-10-05T04:57:46.920Z,
}
`;

exports[`jobTime returns correct time properties for the \`ACCEPTED\` state 2`] = `Object {}`;

exports[`jobTime returns correct time properties for the \`ACCEPTED\` state 3`] = `Object {}`;

exports[`jobTime returns correct time properties for the \`ASSIGNED\` state 1`] = `
Object {
  "from": 2016-10-05T04:57:46.920Z,
}
`;

exports[`jobTime returns correct time properties for the \`ASSIGNED\` state 2`] = `Object {}`;

exports[`jobTime returns correct time properties for the \`ASSIGNED\` state 3`] = `Object {}`;

exports[`jobTime returns correct time properties for the \`BLOCKED_FAILED\` state 1`] = `
Object {
  "from": 2016-10-05T04:57:46.920Z,
}
`;

exports[`jobTime returns correct time properties for the \`BLOCKED_FAILED\` state 2`] = `Object {}`;

exports[`jobTime returns correct time properties for the \`BLOCKED_FAILED\` state 3`] = `Object {}`;

exports[`jobTime returns correct time properties for the \`BLOCKED\` state 1`] = `
Object {
  "from": 2016-10-05T04:57:46.920Z,
}
`;

exports[`jobTime returns correct time properties for the \`BLOCKED\` state 2`] = `Object {}`;

exports[`jobTime returns correct time properties for the \`BLOCKED\` state 3`] = `Object {}`;

exports[`jobTime returns correct time properties for the \`BROKEN\` state 1`] = `
Object {
  "from": 2016-10-05T04:57:46.920Z,
}
`;

exports[`jobTime returns correct time properties for the \`BROKEN\` state 2`] = `Object {}`;

exports[`jobTime returns correct time properties for the \`BROKEN\` state 3`] = `Object {}`;

exports[`jobTime returns correct time properties for the \`CANCELED\` state 1`] = `
Object {
  "from": 2016-10-05T04:57:46.920Z,
  "to": 2016-10-05T04:57:47.000Z,
}
`;

exports[`jobTime returns correct time properties for the \`CANCELED\` state 2`] = `Object {}`;

exports[`jobTime returns correct time properties for the \`CANCELED\` state 3`] = `Object {}`;

exports[`jobTime returns correct time properties for the \`CANCELING\` state 1`] = `
Object {
  "from": 2016-10-05T04:57:46.920Z,
  "to": 2016-10-05T04:57:47.000Z,
}
`;

exports[`jobTime returns correct time properties for the \`CANCELING\` state 2`] = `Object {}`;

exports[`jobTime returns correct time properties for the \`CANCELING\` state 3`] = `Object {}`;

exports[`jobTime returns correct time properties for the \`FINISHED\` state 1`] = `
Object {
  "from": 2016-10-05T04:57:46.920Z,
  "to": 2016-10-05T04:57:47.000Z,
}
`;

exports[`jobTime returns correct time properties for the \`FINISHED\` state 2`] = `Object {}`;

exports[`jobTime returns correct time properties for the \`FINISHED\` state 3`] = `Object {}`;

exports[`jobTime returns correct time properties for the \`PENDING\` state 1`] = `
Object {
  "from": 2016-10-05T04:57:46.920Z,
}
`;

exports[`jobTime returns correct time properties for the \`PENDING\` state 2`] = `Object {}`;

exports[`jobTime returns correct time properties for the \`PENDING\` state 3`] = `Object {}`;

exports[`jobTime returns correct time properties for the \`RUNNING\` state 1`] = `
Object {
  "from": 2016-10-05T04:57:46.920Z,
}
`;

exports[`jobTime returns correct time properties for the \`RUNNING\` state 2`] = `Object {}`;

exports[`jobTime returns correct time properties for the \`RUNNING\` state 3`] = `Object {}`;

exports[`jobTime returns correct time properties for the \`SCHEDULED\` state 1`] = `
Object {
  "from": 2016-10-05T04:57:46.920Z,
}
`;

exports[`jobTime returns correct time properties for the \`SCHEDULED\` state 2`] = `Object {}`;

exports[`jobTime returns correct time properties for the \`SCHEDULED\` state 3`] = `Object {}`;

exports[`jobTime returns correct time properties for the \`SKIPPED\` state 1`] = `
Object {
  "from": 2016-10-05T04:57:46.920Z,
}
`;

exports[`jobTime returns correct time properties for the \`SKIPPED\` state 2`] = `Object {}`;

exports[`jobTime returns correct time properties for the \`SKIPPED\` state 3`] = `Object {}`;

exports[`jobTime returns correct time properties for the \`TIMED_OUT\` state 1`] = `
Object {
  "from": 2016-10-05T04:57:46.920Z,
  "to": 2016-10-05T05:14:26.920Z,
}
`;

exports[`jobTime returns correct time properties for the \`TIMED_OUT\` state 2`] = `Object {}`;

exports[`jobTime returns correct time properties for the \`TIMED_OUT\` state 3`] = `Object {}`;

exports[`jobTime returns correct time properties for the \`TIMING_OUT\` state 1`] = `
Object {
  "from": 2016-10-05T04:57:46.920Z,
  "to": 2016-10-05T05:14:26.920Z,
}
`;

exports[`jobTime returns correct time properties for the \`TIMING_OUT\` state 2`] = `Object {}`;

exports[`jobTime returns correct time properties for the \`TIMING_OUT\` state 3`] = `Object {}`;

exports[`jobTime returns correct time properties for the \`UNBLOCKED_FAILED\` state 1`] = `
Object {
  "from": 2016-10-05T04:57:46.920Z,
}
`;

exports[`jobTime returns correct time properties for the \`UNBLOCKED_FAILED\` state 2`] = `Object {}`;

exports[`jobTime returns correct time properties for the \`UNBLOCKED_FAILED\` state 3`] = `Object {}`;

exports[`jobTime returns correct time properties for the \`UNBLOCKED\` state 1`] = `
Object {
  "from": 2016-10-05T04:57:46.920Z,
}
`;

exports[`jobTime returns correct time properties for the \`UNBLOCKED\` state 2`] = `Object {}`;

exports[`jobTime returns correct time properties for the \`UNBLOCKED\` state 3`] = `Object {}`;

exports[`jobTime returns correct time properties for the \`WAITING_FAILED\` state 1`] = `
Object {
  "from": 2016-10-05T04:57:46.920Z,
}
`;

exports[`jobTime returns correct time properties for the \`WAITING_FAILED\` state 2`] = `Object {}`;

exports[`jobTime returns correct time properties for the \`WAITING_FAILED\` state 3`] = `Object {}`;

exports[`jobTime returns correct time properties for the \`WAITING\` state 1`] = `
Object {
  "from": 2016-10-05T04:57:46.920Z,
}
`;

exports[`jobTime returns correct time properties for the \`WAITING\` state 2`] = `Object {}`;

exports[`jobTime returns correct time properties for the \`WAITING\` state 3`] = `Object {}`;

exports[`jobWaitTime returns correct time properties for the \`ACCEPTED\` state 1`] = `
Object {
  "from": "2017-10-29T00:32:23.479Z",
  "to": "2017-10-29T00:34:37.000Z",
}
`;

exports[`jobWaitTime returns correct time properties for the \`ACCEPTED\` state 2`] = `
Object {
  "from": "2017-10-29T00:32:23.479Z",
}
`;

exports[`jobWaitTime returns correct time properties for the \`ASSIGNED\` state 1`] = `
Object {
  "from": "2017-10-29T00:32:23.479Z",
  "to": "2017-10-29T00:34:37.000Z",
}
`;

exports[`jobWaitTime returns correct time properties for the \`ASSIGNED\` state 2`] = `
Object {
  "from": "2017-10-29T00:32:23.479Z",
}
`;

exports[`jobWaitTime returns correct time properties for the \`BLOCKED_FAILED\` state 1`] = `
Object {
  "from": "2017-10-29T00:32:23.479Z",
  "to": "2017-10-29T00:34:37.000Z",
}
`;

exports[`jobWaitTime returns correct time properties for the \`BLOCKED_FAILED\` state 2`] = `
Object {
  "from": "2017-10-29T00:32:23.479Z",
}
`;

exports[`jobWaitTime returns correct time properties for the \`BLOCKED\` state 1`] = `Object {}`;

exports[`jobWaitTime returns correct time properties for the \`BLOCKED\` state 2`] = `Object {}`;

exports[`jobWaitTime returns correct time properties for the \`BROKEN\` state 1`] = `
Object {
  "from": "2017-10-29T00:32:23.479Z",
  "to": "2017-10-29T00:34:37.000Z",
}
`;

exports[`jobWaitTime returns correct time properties for the \`BROKEN\` state 2`] = `
Object {
  "from": "2017-10-29T00:32:23.479Z",
}
`;

exports[`jobWaitTime returns correct time properties for the \`CANCELED\` state 1`] = `
Object {
  "from": "2017-10-29T00:32:23.479Z",
  "to": "2017-10-29T00:34:37.000Z",
}
`;

exports[`jobWaitTime returns correct time properties for the \`CANCELED\` state 2`] = `
Object {
  "from": "2017-10-29T00:32:23.479Z",
}
`;

exports[`jobWaitTime returns correct time properties for the \`CANCELING\` state 1`] = `
Object {
  "from": "2017-10-29T00:32:23.479Z",
  "to": "2017-10-29T00:34:37.000Z",
}
`;

exports[`jobWaitTime returns correct time properties for the \`CANCELING\` state 2`] = `
Object {
  "from": "2017-10-29T00:32:23.479Z",
}
`;

exports[`jobWaitTime returns correct time properties for the \`FINISHED\` state 1`] = `
Object {
  "from": "2017-10-29T00:32:23.479Z",
  "to": "2017-10-29T00:34:37.000Z",
}
`;

exports[`jobWaitTime returns correct time properties for the \`FINISHED\` state 2`] = `
Object {
  "from": "2017-10-29T00:32:23.479Z",
}
`;

exports[`jobWaitTime returns correct time properties for the \`PENDING\` state 1`] = `
Object {
  "from": "2017-10-29T00:32:23.479Z",
  "to": "2017-10-29T00:34:37.000Z",
}
`;

exports[`jobWaitTime returns correct time properties for the \`PENDING\` state 2`] = `
Object {
  "from": "2017-10-29T00:32:23.479Z",
}
`;

exports[`jobWaitTime returns correct time properties for the \`RUNNING\` state 1`] = `
Object {
  "from": "2017-10-29T00:32:23.479Z",
  "to": "2017-10-29T00:34:37.000Z",
}
`;

exports[`jobWaitTime returns correct time properties for the \`RUNNING\` state 2`] = `
Object {
  "from": "2017-10-29T00:32:23.479Z",
}
`;

exports[`jobWaitTime returns correct time properties for the \`SCHEDULED\` state 1`] = `
Object {
  "from": "2017-10-29T00:32:23.479Z",
  "to": "2017-10-29T00:34:37.000Z",
}
`;

exports[`jobWaitTime returns correct time properties for the \`SCHEDULED\` state 2`] = `
Object {
  "from": "2017-10-29T00:32:23.479Z",
}
`;

exports[`jobWaitTime returns correct time properties for the \`SKIPPED\` state 1`] = `
Object {
  "from": "2017-10-29T00:32:23.479Z",
  "to": "2017-10-29T00:34:37.000Z",
}
`;

exports[`jobWaitTime returns correct time properties for the \`SKIPPED\` state 2`] = `
Object {
  "from": "2017-10-29T00:32:23.479Z",
}
`;

exports[`jobWaitTime returns correct time properties for the \`TIMED_OUT\` state 1`] = `
Object {
  "from": "2017-10-29T00:32:23.479Z",
  "to": "2017-10-29T00:34:37.000Z",
}
`;

exports[`jobWaitTime returns correct time properties for the \`TIMED_OUT\` state 2`] = `
Object {
  "from": "2017-10-29T00:32:23.479Z",
}
`;

exports[`jobWaitTime returns correct time properties for the \`TIMING_OUT\` state 1`] = `
Object {
  "from": "2017-10-29T00:32:23.479Z",
  "to": "2017-10-29T00:34:37.000Z",
}
`;

exports[`jobWaitTime returns correct time properties for the \`TIMING_OUT\` state 2`] = `
Object {
  "from": "2017-10-29T00:32:23.479Z",
}
`;

exports[`jobWaitTime returns correct time properties for the \`UNBLOCKED_FAILED\` state 1`] = `
Object {
  "from": "2017-10-29T00:32:23.479Z",
  "to": "2017-10-29T00:34:37.000Z",
}
`;

exports[`jobWaitTime returns correct time properties for the \`UNBLOCKED_FAILED\` state 2`] = `
Object {
  "from": "2017-10-29T00:32:23.479Z",
}
`;

exports[`jobWaitTime returns correct time properties for the \`UNBLOCKED\` state 1`] = `
Object {
  "from": "2017-10-29T00:32:23.479Z",
  "to": "2017-10-29T00:34:37.000Z",
}
`;

exports[`jobWaitTime returns correct time properties for the \`UNBLOCKED\` state 2`] = `
Object {
  "from": "2017-10-29T00:32:23.479Z",
}
`;

exports[`jobWaitTime returns correct time properties for the \`WAITING_FAILED\` state 1`] = `
Object {
  "from": "2017-10-29T00:32:23.479Z",
  "to": "2017-10-29T00:34:37.000Z",
}
`;

exports[`jobWaitTime returns correct time properties for the \`WAITING_FAILED\` state 2`] = `
Object {
  "from": "2017-10-29T00:32:23.479Z",
}
`;

exports[`jobWaitTime returns correct time properties for the \`WAITING\` state 1`] = `
Object {
  "from": "2017-10-29T00:32:23.479Z",
  "to": "2017-10-29T00:34:37.000Z",
}
`;

exports[`jobWaitTime returns correct time properties for the \`WAITING\` state 2`] = `
Object {
  "from": "2017-10-29T00:32:23.479Z",
}
`;
